numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1770,23 +1783,72 @@ array_iter(PyArrayObject *arr)
     return PySeqIter_New((PyObject *)arr);
 }
 
-static PyObject *
-array_alloc(PyTypeObject *type, Py_ssize_t NPY_UNUSED(nitems))
+
+/*
+ * Impelement traversal for cyclic garbage collection. There are two main
+ * cases. Arrays that do not own their own data, must VISIT their base.
+ * Arrays which hold the actual data own the references and have to visit
+ * all items.
+ */
+static int
+array_traverse(PyArrayObject *self, visitproc visit, void *arg)
 {
-    /* nitems will always be 0 */
-    PyObject *obj = PyObject_Malloc(type->tp_basicsize);
-    PyObject_Init(obj, type);
-    return obj;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    /*
+     * make sure we don't traverse the array before it is fully initialized,
+     * or if it doesn't have any objects
+     */
+    if (!descr) {
+        return 0;
```

Comment:
According to cov, the `!descr` path is never hit. Is it reasonable to remove the check?

Reply:
Changed it `PyArray_DATA(self) == NULL` for now, the issue is that allocation adds the object to the GC tracker too early on in a sense. It probably cannot happen that the traverse ever kicks in, since we are busy within C if it fails at this point. But that is what this seems to have been for (the change was by @J-Sand, maybe he has a quick note on it).

It is likely that this is just a band-aid solution, and the correct thing is to ensure that tracking is only enabled after everything is initialized. This might be a major rough edge :/ thought this was further along.